src/main/java/net/minecraftforge/common/data/GrindingRecipeBuilder.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.data;
+
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import com.google.gson.JsonObject;
+
+import net.minecraft.advancements.Advancement;
+import net.minecraft.advancements.AdvancementRewards;
+import net.minecraft.advancements.ICriterionInstance;
+import net.minecraft.advancements.IRequirementsStrategy;
+import net.minecraft.advancements.criterion.RecipeUnlockedTrigger;
+import net.minecraft.data.IFinishedRecipe;
+import net.minecraft.item.Item;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.registry.Registry;
+import net.minecraftforge.common.crafting.GrindingRecipe;
+
+/**
+ * Class for building a grinding recipe. Takes 2 ingredients as crafting materials and a item as result.
+ */
+public class GrindingRecipeBuilder
+{
+    private final Ingredient base;
+    private final Ingredient addition;
+    private final Item result;
+    private final Advancement.Builder advancement = Advancement.Builder.advancement();
+    private final IRecipeSerializer<?> type;
+    
+    public GrindingRecipeBuilder(IRecipeSerializer<?> type, Ingredient base, Ingredient addition, Item result)
```

Comment:
Shouldn't be `Item` as data type, should serialize to `ItemStack` since that is supported.